File: lambda/modules/classtime.js
class ClassTime{
    constructor(days, time, instructor, location){
        this.days = days;
        this.instructor = instructor;
        this.location = location;
        this.building = location.slice(0, location.lastIndexOf(" ")); //The building name comes before the last space

        //Create numbers in 24-hour time to represent the time of day. A hyphen separates the start and end times in the string
        this.startTime = ClassTime.createTimeOfDay(time.slice(0, time.indexOf("-") - 1)); //Account for the space before the hyphen
        this.endTime = ClassTime.createTimeOfDay(time.slice(time.indexOf("-") + 2)); //Start index is inclusive so +2 instead of +1
    }

    //Converts a string into a Date object representing just the time of day
    static createTimeOfDay(time){
        var hour = time.contains("pm") ? parseInt(time.slice(0, 2)) + 12 : parseInt(time.slice(0, 2)); //Convert hours to military time
        var minute = parseInt(time.slice(3, 5)); //Extract the minutes from the string

        //Set the time of day using the hour and minute values we extracted. We need to parse the number back out of a string to ensure it's read properly
        //The return type must also be a number since it is used for comparisons in the functions below
        return parseInt(`${hour}${minute}`);
    }

    //Takes 2 ClassTime objects and returns whether they have any days in common
    static onSameDay(class1, class2){
        var combined = class1.days + class2.days; //Add all of the characters together
        return (/([a-zA-Z]).*?\1/).test(combined); //Regex will match any duplicates, therefore the 2 classes share days
    }

    //Takes 2 ClassTime objects and returns whether they have a time conflict
    static hasTimeConflict(class1, class2){
        if (!ClassTime.onSameDay(class1, class2)){ //If both classes are not on the same day, we don't need to check the times
            return false;
        }

        //See if the start or end time for class1 falls in between class2's start and end times
        var startConflict = class1.startTime >= class2.startTime && class1.startTime <= class2.endTime;
        var endConflict = class1.endTime >= class2.startTime && class1.endTime <= class2.endTime;

        return startConflict || endConflict;
    }
}

module.exports = ClassTime;

File: original C# files/ClassTime.cs
using System;
using System.Linq;

namespace ScheduleMaker
{
    public class ClassTime
    {
        //FIELDS
        private string location, instructor, days, building;
        private DateTime startTime, endTime; //DateTime allows for comparing times

        //CONSTRUCTORS
        public ClassTime()
        {

        }

        public ClassTime(string days, string time, string instructor, string location)
        {
            this.location = location;
            //this.building = location.Remove(location.LastIndexOf(' ')); //removes the room number and leaves just the building name. Used by Laith for calculating distance
            this.instructor = instructor;
            this.days = days;
            this.startTime = DateTime.Parse(time.Remove(time.IndexOf(" -"))); //removes everything after the dash
            this.endTime = DateTime.Parse(time.Substring(time.IndexOf("- ")+1)); //starts after the dash. Index offset by 1 to make it start after the space
        }

        //ACCESSORS
        public DateTime StartTime
        {
            get { return this.startTime; }
            set { this.startTime = value; }
        }

        public DateTime EndTime
        {
            get { return this.endTime; }
            set { this.endTime = value; }
        }

        public string Location
        {
            get { return this.location; }
            set { this.location = value; }
        }

        public string Days
        {
            get { return this.days; }
            set { this.days = value; }
        }

        public string Instructor
        {
            get { return this.instructor; }
            set { this.instructor = value; }
        }

        public string Building
        {
            get { return this.building; }
            set { this.building = value; }
        }

        public String GetStartTime() //should only be used for displaying the data to the user. StartTime accessor should be used for time comparisons
        {
            return this.startTime.ToShortTimeString();
        }

        public String GetEndTime() //should only be used for displaying the data to the user. EndTime accessor should be used for time comparisons
        {
            return this.endTime.ToShortTimeString();
        }


        //METHODS
        public bool OnSameDay(ClassTime AClass)
        {
            foreach (char chr in this.days) //test all chars in the days shorthand
            {
                if (AClass.Days.Contains(chr)) 
                {
                    return true; //if AClass and this have any day(s) in common
                }
            }
            return false; //if it passes the entire foreach loop, then AClass and this have no days in common
        }

        public bool HasTimeConflict(ClassTime AClass)
        {
            if ((AClass.StartTime.TimeOfDay >= this.startTime.TimeOfDay) && (AClass.StartTime.TimeOfDay <= this.endTime.TimeOfDay)) 
            {
                return true; //if the start time for AClass is in between the start and end times for this, then there is a time conflict
            }
            else if ((AClass.EndTime.TimeOfDay >= this.startTime.TimeOfDay) && (AClass.EndTime.TimeOfDay <= this.endTime.TimeOfDay))
            {
                return true; //if the end time for AClass is in between the start and end times for this, then there is a time conflict
            }
            else return false; //only if the start and end times don't conflict is there no time conflict
        }
    }
}
